packages/react-dom/src/__tests__/ReactTestUtilsAct-test.js

```diff
@@ -31,476 +55,502 @@ describe('ReactTestUtils.act()', () => {
-              calledCounter = val;
-            }}
-          />,
-          container,
-        );
-      });
-      const button = document.getElementById('button');
-      function click() {
-        button.dispatchEvent(new MouseEvent('click', {bubbles: true}));
-      }
-
-      act(() => {
-        click();
-        click();
-        click();
-      });
-      expect(calledCounter).toBe(3);
-      act(click);
-      expect(calledCounter).toBe(4);
-      act(click);
-      expect(calledCounter).toBe(5);
-      expect(button.innerHTML).toBe('5');
+    it("should warn when the scheduler isn't mocked", () => {
+      expect(() => act(() => {})).toWarnDev([
+        'Starting from React v17, the "scheduler" module will need to be mocked',
+      ]);
     });
+  });
 
-    it('should flush effects recursively', () => {
-      function App() {
-        let [ctr, setCtr] = React.useState(0);
-        React.useEffect(() => {
-          if (ctr < 5) {
-            setCtr(x => x + 1);
+  [false, true].forEach(concurrent => {
+    isConcurrentMode = concurrent;
+    describe(isConcurrentMode ? 'concurrent mode' : 'heritage mode', () => {
+      describe('sync', () => {
+        it('can use act to flush effects', () => {
+          function App(props) {
+            React.useEffect(props.callback);
+            return null;
           }
-        });
-        return ctr;
-      }
-
-      act(() => {
-        ReactDOM.render(<App />, container);
-      });
 
-      expect(container.innerHTML).toBe('5');
-    });
-
-    it('detects setState being called outside of act(...)', () => {
-      let setValue = null;
-      function App() {
-        let [value, _setValue] = React.useState(0);
-        setValue = _setValue;
-        return (
-          <button id="button" onClick={() => setValue(2)}>
-            {value}
-          </button>
-        );
-      }
-      let button;
-      act(() => {
-        ReactDOM.render(<App />, container);
-        button = container.querySelector('#button');
-        button.dispatchEvent(new MouseEvent('click', {bubbles: true}));
-      });
-      expect(button.innerHTML).toBe('2');
-      expect(() => setValue(1)).toWarnDev([
-        'An update to App inside a test was not wrapped in act(...).',
-      ]);
-    });
-    describe('fake timers', () => {
-      beforeEach(() => {
-        jest.useFakeTimers();
-      });
-      afterEach(() => {
-        jest.useRealTimers();
-      });
-      it('lets a ticker update', () => {
-        function App() {
-          let [toggle, setToggle] = React.useState(0);
-          React.useEffect(() => {
-            let timeout = setTimeout(() => {
-              setToggle(1);
-            }, 200);
-            return () => clearTimeout(timeout);
-          }, []);
-          return toggle;
-        }
+          let calledLog = [];
+          act(() => {
+            render(
+              <App
+                callback={() => {
+                  calledLog.push(calledLog.length);
```

Comment:
Nit: Why this prop indirection? Why does it log the length? Might seem like a silly nit to pick but little stuff like this makes it hard for me to infer which parts of the test are important and which are incidental.

Reply:
now that we have a mock scheduler, I should just use .yieldValue and co